src/librustc/traits/error_reporting/mod.rs

```diff
@@ -917,14 +917,27 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             // already reported in the query
             ConstEvalFailure(err) => {
                 if let ErrorHandled::TooGeneric = err {
-                    // Silence this error, as it can be produced during intermediate steps
-                    // when a constant is not yet able to be evaluated (but will be later).
-                    return;
+                    // In this instance, we have a const expression containing an unevaluated
+                    // generic parameter. We have no idea whether this expression is valid or
+                    // not (e.g. it might result in an error), but we don't want to just assume
+                    // that it's okay, because that might result in post-monomorphisation time
+                    // errors. The onus is really on the caller to provide values that it can
+                    // prove are well-formed.
+                    let mut err = self.tcx.sess.struct_span_err(
+                        span,
+                        "cannot determine that constant is always well-formed",
```

Comment:
We probably shouldn't talk about WF in diagnostics.

The way I was initially approaching this (see variant names, like `ConstEvaluatable` in `Predicate` and `ConstEvalFailure` here) would suggest something more like:
> constant expression depends on generic parameters

followed by explaining that it may fail to evaluate depending on the choice of generics.

Reply:
Oh and you can bypass the problem of explaining to the user how to make it work, by linking to a `rust-lang/rust` issue (ideally a new one without any preexisting discussion) titled something like "Constant expressions using generic params require weird `where` clauses." with a "canonical" example.

Note that the "canonical" example doesn't work yet, because `ConstEvaluatable` requirements from bounds aren't implied in the body (maybe they need to be added to `ParamEnv`, and then looked up in there?).

I'm not even sure we can (ab)use the `where TyAlias<T, {N}>:,` pattern like that, usually WF semantics would require that the definition that `where` is attached to proves the WF, instead of deferring that to users of that definition.
I would want a @rust-lang/wg-traits confirmation that we want to implement that behavior (as it's not limited to const generics, it applies to any constant expressions using generic parameters, e.g. `[u8; size_of::<T>()]`, so it might impact stable).